### PR TITLE
Download

### PR DESCRIPTION
**Pull request by otaviobp**
Pull request with some fixes and some preparation to the fix of #298.

Please, consider a careful review of "5322aff: download: remove mutex lock in single thread code" because I don't know the reason the mutex lock was added to that file and I may have missed something.

Another point I'd like to be carefully reviewed is the "curl: Set fallback CA paths in swupd_curl_init()" commit. Is there any reason we needed to check the connection using swupd_curl_check_network() in a moment different from when we called swupd_curl_init()?

### CODE REVIEW

**Comment by phmccarty** [comment]
> Please, consider a careful review of "5322aff: download: remove mutex lock in single thread code" because I don't know the reason the mutex lock was added to that file and I may have missed something.

You didn't miss anything. This part of the code used to be multithreaded, but it predates the Github git history for the repo. So, removing the mutex locks is safe.

File: src/download.c [hunk 1]
```diff
@@ -58,9 +58,12 @@ static struct list *failed = NULL;
  * traversing a manifest, doing hash comparisons, and (re)staging any files
  * whose hash miscompares.
  *
- * file->hash[0] acts as index into the arrays
+ * file->hash[0:1] acts as index into the arrays
  */
 #define SWUPD_CURL_HASH_BUCKETS 256
+#define HASH_VALUE(h) (h >= 'a' ? h - 'a' + 10 : h - '0')
+#define HASH_TO_KEY(hash) (HASH_VALUE(hash[0]) << 4 | HASH_VALUE(hash[1]))
```
Comment (matthewrsj) [inline src/download.c]: Maybe I'm being dense, but can you explain some of the magic here? Why are you shifting the first hash value left 4? Why add 10 to h - 'a' ?

File: src/curl.c [hunk 2]
```diff
@@ -187,6 +148,10 @@ void swupd_curl_deinit(void)
 		curl_easy_cleanup(curl);
 	}
 	curl = NULL;
+
+	free(capath);
```
Comment (matthewrsj) [inline src/curl.c]: Can't you just `free_string(&capath);`?

File: test/functional/verify/latest-missing/test.bats [hunk 3]
```diff
@@ -5,7 +5,7 @@ load "../../swupdlib"
 @test "verify install using latest with missing version file on server" {
   run sudo sh -c "$SWUPD verify $SWUPD_OPTS --install -m latest"
 
-  check_lines "$output"
```
Comment (matthewrsj) [inline test/functional/verify/latest-missing/test.bats]: Why aren't we checking the output any more? Are we losing the error message?

File: src/download.c (hunk 1 above)
Comment (icarus-sparry) [reply src/download.c]: HASH_VALUE converts a character to its hexadecimal value (0 to 15 assuming that the input was from the set 0123456789abcdef ).
HASH_TO_KEY converts two characters to a value in the range 0 to 255.

File: src/curl.c (hunk 2 above)
Comment (otaviobp) [reply src/curl.c]: Yes. I should have used it.

File: test/functional/verify/latest-missing/test.bats (hunk 3 above)
Comment (otaviobp) [reply test/functional/verify/latest-missing/test.bats]: Good question.

This test is checking the behavior of swupd when the latest file was missing in the server. Before this patch CA Paths weren't used when downloading the latest file, so the error messages have changed. The problem is that the output will be different if the code is compiled with or without CA paths and looked enough. I haven't realized though that this is easy to fix using reg exp. I'll fix that and check the output correctly.

**Comment by otaviobp** [comment]
On Tue, 2018-07-03 at 08:31 -0700, Matthew Johnson wrote:
> @matthewrsj commented on this pull request.
> curl: Aborting on invalid parameter is a bit too much: +1 on the
> change, but the commit message is confusing. This isn't an invalid
> parameter, this indicates that curl hasn't been initialized (it isn't
> in the parameter list).

My bad. I'll fix the commit message

File: src/download.c (hunk 1 above)
Comment (otaviobp) [reply src/download.c]: Exactly. I'll comment this macro.

**Comment by otaviobp** [comment]
Sent new version addressing the suggestions

File: test/functional/verify/latest-missing/lines-checked [hunk 4]
```diff
@@ -1,2 +1,2 @@
-Unable to get latest version for install
-Error: Fix did not fully succeed
+Curl error: (37) Couldn't read a file:// file
+Failed verify initialization, exiting now.
```
Comment (phmccarty) [inline test/functional/verify/latest-missing/lines-checked]: Is this an expected change in test output?

Comment (otaviobp) [reply test/functional/verify/latest-missing/lines-checked]: yes

**Pull request closed without merging**